tests_tf/test_attacks.py

```diff
@@ -974,6 +891,40 @@ def test_generate_np_gives_clipped_adversarial_examples(self):
     self.assertLess(-0.201, np.min(x_adv))
     self.assertLess(np.max(x_adv), .301)
 
+
+  def test_generate_np_does_not_cache_graph_computation_for_nb_iter(self):
+    x_val = np.random.rand(100, 2)
+    x_val = np.array(x_val, dtype=np.float32)
+
+    x_adv = self.attack.generate_np(x_val, eps=1.0, ord=np.inf,
+                                    clip_min=-5.0, clip_max=5.0,
+                                    nb_iter=10)
+
+    orig_labs = np.argmax(self.sess.run(self.model.get_logits(x_val)), axis=1)
+    new_labs = np.argmax(self.sess.run(self.model.get_logits(x_adv)), axis=1)
+    self.assertTrue(np.mean(orig_labs == new_labs) < 0.1)
+
+    ok = [False]
+    old_grads = tf.gradients
+
+    def fn(*x, **y):
+      ok[0] = True
+      return old_grads(*x, **y)
+
+    tf.gradients = fn
```

Comment:
If the test crashes inside `generate_np` or something like, will all the other tests get stuck using the wrong gradients?

It might be better to test this by running the attack on a model whose fprop method contains the instrumentation, rather than by changing a shared library.

Another option might be to use a `try`-`finally` construction to make sure that the replacement always happens.

Reply:
This code all already existed. I was just moving it.